integration_tests/config-refresh_test.go

```diff
@@ -0,0 +1,166 @@
+//
+// Copyright (c) 2023 Red Hat, Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integration_tests
+
+import (
+	"context"
+	"fmt"
+	"redhat-developer/red-hat-developer-hub-operator/pkg/utils"
+	"strings"
+	"time"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	corev1 "k8s.io/api/core/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+
+	"redhat-developer/red-hat-developer-hub-operator/pkg/model"
+
+	bsv1alpha1 "redhat-developer/red-hat-developer-hub-operator/api/v1alpha1"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = When("create backstage with external configuration", func() {
+
+	var (
+		ctx context.Context
+		ns  string
+	)
+
+	BeforeEach(func() {
+		ctx = context.Background()
+		ns = createNamespace(ctx)
+	})
+
+	AfterEach(func() {
+		deleteNamespace(ctx, ns)
+	})
+
+	It("refresh config", func() {
```

Comment:
This one would test if builder.WithPredicates works correct with LabelSelectorPredicate (does not call enqueue in this case). I would prefer to rely on K8s and avoid testing it's behaviour with costly tests (requiring real (i. e. not envtest) cluster and so quite long)

Reply:
> This one would test if builder.WithPredicates works correct with LabelSelectorPredicate (does not call enqueue in this case). I would prefer to rely on K8s

Well, this is the case now, but how can I guarantee that this will continue to behave this way with future code changes? The thing I am mostly concerned about is: if I remove the code that handles the env var that controls the auto-sync mechanism, the expectations would no longer be correct, but the current tests would still pass. So we'd better catch this before the customer does.
Could this be covered with unit tests instead?

> avoid testing it's behaviour with costly tests (requiring real (i. e. not envtest) cluster and so quite long)

Also, I think that was the point of the nightly E2E tests.. It should be fine even if they take long (with a reasonable timeout of course), as long as they allow us to catch potential regressions earlier.

Reply:
> Well, this is the case now, but how can I guarantee that this will continue to behave this way with future code changes? The thing I am mostly concerned about is: if I remove the code that handles the env var that controls the auto-sync mechanism, the expectations would no longer be correct, but the current tests would still pass. So we'd better catch this before the customer does.
> Could this be covered with unit tests instead?

I can not get your point, sorry.
Like I said I rely on K8s/go functionality, i.e. no label - no call, no annotation fitting backstage name - no call, no changes in configuration hash => no Pod recreating (deployment is identical), I do not test it here, the same as we do not usually test whether K8s's GET by name or Go's if work correct.  I am interested in testing functionality I added - i.e. when it all triggered - configuration changed (for other case we have other tests) ... It is a complex test and I definitely can not do it with unit or even envtest because it requires working deployment/pod controllers.
"if I remove the code that handles the env var that controls the auto-sync mechanism", FYI:
- env var does not control auto-sync mechanism, it is just opposite
- I guess, you can remove the code which handles re-creating env vars only with the piece of Pod controller  ;)

Reply:
I understand, but the code changes introduce 2 paths: one (now default behavior) which is to always recreate the pod upon config changes, and another one (which is a valid and supported use case) is to not do it. Right now, only the former is tested.
Without going too deep into a real test, I was thinking of simply adding a test case to `cr-config_test.go`, to ensure that the `rhdh.redhat.com/ext-config-sync` label value can be `false` depending on the operator environment (like https://github.com/gazarenkov/janus-idp-operator/blob/refresh_config/integration_tests/cr-config_test.go#L187). Does that make sense?

Reply:
Well, there is one place where this label is created/updated by operator (updating the CM/Secret):

````
		obj.GetLabels()[model.ExtConfigSyncLabel] = strconv.FormatBool(autoSync)
		obj.GetAnnotations()[model.BackstageNameAnnotation] = backstageName
		if err := r.Update(ctx, obj); err != nil {
````
Where autoSync is read from env variable or use 'true' by default.

```
        autoSync := true
	autoSyncStr, ok := os.LookupEnv(AutoSyncEnvVar)
	if ok {
		autoSync, _ = strconv.ParseBool(autoSyncStr)
	}

```
Did I understand you correct - you want me to make a test which 
- sets env var to 'false'
- reconsiles CR with ext CM/Secret 
- and check if CM/Secret's annotation is 'false' too?
(yes, I can do that if it helps)

Reply:
Yes, that's exactly the path that I thought would make sense to cover: the condition that leads to setting this annotation to `false` (which should in turn prevent recreating the pod upon changes IIUC). If it makes sense of course :D
Thanks!

Reply:
test added